### PR TITLE
feat(ptx): exact-shape FP32 row-softmax blueprint (#840)

### LINKED ISSUE #840: [PTX] Port softmax, log-softmax, masking, log-sum-exp, and backward fusions
Parent epic: #833

Foundation PR: #832

## Operation scope

Implement dense/row/axis FP16/BF16/FP32 softmax, log-softmax, softmax backward, log-sum-exp and backward, triangular/causal masking, masked fill and backward, probability normalization, sparsemax, Taylor softmax, spherical softmax, and other attention-independent normalization-of-logits paths.

Fuse scale/bias/mask/softmax and eligible consumer stages, use online/warp reductions, and avoid global max/sum/probability intermediates. Cross-entropy ownership remains in the dedicated PTX loss child, with a shared softmax ABI where appropriate.

### Inventory and semantics

- [ ] Enumerate every public AiDotNet operation and existing CUDA/NVRTC/library-backed kernel in this scope; the issue is not complete while any scoped entry point is unassigned.
- [ ] Define the intended fusion boundaries, supported forward/backward modes, dtypes, logical shapes, physical layouts, numerical modes, and architecture matrix.

### PTX blueprint and ABI

- [ ] Implement the operation as direct Driver-API-loaded PTX, ported from the existing kernel or written from scratch; CUDA C/NVRTC hidden behind a new wrapper does not count.
- [ ] Give every admitted specialization a versioned `DirectPtxKernelBlueprint` with PTX SHA-256, GPU UUID/SM/driver identity, block geometry, resource budget, and occupancy.
- [ ] Declare formal tensor contracts for logical/physical extents, dtype, layout identity, alignment, byte offset, allocation/view extent, access, aliasing, and padding.
- [ ] Validate the contract once before dispatch. Admitted PTX must contain no runtime stride/layout checks and no generic dynamic-shape branches.
- [ ] Keep Ampere, Ada, Hopper, and Blackwell specializations separate. Unmeasured architectures and semantics must fail closed to the established backend with an exact reason.

### Fused dataflow

- [ ] Document every global-memory read/write and identify which intermediates remain in registers or shared memory.
- [ ] Eliminate avoidable global intermediates and repeated reads; use `cp.async`, Tensor Cores, warp reductions, vectorized transactions, or architecture-specific instructions wherever they are appropriate and measurably beneficial.
- [ ] Prove required alignment/coalescing and shared-memory banking for every physical layout; semantic padding may not be introduced silently.

### Correctness and runtime behavior

- [ ] Match a high-precision oracle and current AiDotNet on a wide shape/dtype/edge-case matrix, including advertised forward and backward behavior.
- [ ] Test every eligibility rejection, fallback reason, alias rule, numerical corner case, and deterministic mode.
- [ ] Provide bounded GPU/driver/shape/semantics-keyed autotuning and bounded module/plan caches with deterministic cleanup.
- [ ] Provide prewarm and CUDA graph-capture tests. Capture may not JIT, tune, allocate, perform file I/O, or evict a live module.

### Resource and performance proof

- [ ] Reject any JIT result with nonzero local bytes. Record registers, static/dynamic shared memory, local bytes, max threads, PTX/binary version, occupancy, hash, and JIT log.
- [ ] Attach Nsight Compute evidence showing zero executed register spills, local loads, local stores, and local spilling requests for every promoted specialization.
- [ ] Benchmark resident, apples-to-apples NVIDIA paths against current AiDotNet and the strongest relevant cuBLAS/cuBLASLt/cuDNN/cuSPARSE/cuFFT/PyTorch/other CUDA implementation. Do not compare CPU results to GPU results.
- [ ] Report median, P95, P99, mean, GFLOPS/TFLOPS or GB/s, managed bytes/call, temporary device bytes, maximum numerical error, registers, shared/local bytes, occupancy, and the complete environment fingerprint.
- [ ] Use at least 30 warmups plus 101 samples and three independent clean runs.
- [ ] Pass the production gate: at least 1.10x median speedup over the strongest eligible competitor, candidate P95 no worse than competitor P95 +10%, zero hot managed allocation, zero avoidable temporary VRAM, operation-specific error tolerance, and zero local/spill evidence.

### Integration

- [ ] Ship disabled-by-default/per-kernel rollout gates, established-backend fallback, disposal coverage, documentation, and focused tests.
- [ ] Update the repository PTX coverage inventory and blueprint document. The implementation PR must link this child and include the full benchmark/evidence table.

### PR DESCRIPTION
**Pull request by ooples**
## Merge decision and competitive evidence

| Gate | Current result |
|---|---|
| Live head | `3dd12e0c` |
| Merge status | **HOLD** — draft; no clean three-run p15 matrix and the external semantic contract is not exact. |
| Scope | Four admitted exact-shape FP32 row-softmax buckets; production promotion remains closed. |
| Measurement protocol | **p15** is now on this branch: calibrated equal-duration paired batches, unchanged 5% spread gate, and 20% host-admission ceiling with named PID diagnostics. |
| Foundation verification | Release benchmark build passed; identity/calibration tests passed **6/6** on .NET 10; the worktree was clean before publication. |
| Comparison design | Established AiDotNet plus PyTorch eager, CUDA graph, max-autotune compile, and compile+graph on resident NVIDIA buffers. |
| Semantic disclosure | Direct PTX uses approximate `ex2.approx.ftz`/`rcp.approx.ftz` under an inference tolerance, while `torch.softmax` uses a more precise exponential path. Any speed claim must be labelled tolerance-equivalent, not bit-exact. |
| Competitive qualification | **No current qualifying three-run p15 result; 0/4 promoted shapes.** |

**Required to clear HOLD:** either match the precise competitor contract or explicitly approve and compare the same fast-math/tolerance contract, then run all shapes in three independent host-clean p15 processes against the strongest symmetric graph lane with >=1.10x median, P95 within +10%, correctness, allocation/temporary-VRAM, and zero-spill proof.

> [!IMPORTANT]
> **Scope audit: incomplete.** This draft currently implements only part of its linked child issue. Planned/manifest-only cells do not satisfy acceptance. The PR will remain draft until every operation listed in the issue has a production direct-PTX implementation and integration; only GPU execution, benchmark values, and Nsight evidence may remain pending.

Closes #840.

Establishes the direct-PTX softmax assembly line ([Epic #833](https://github.com/ooples/AiDotNet.Tensors/issues/833)) following the fused-attention championship blueprint pattern (#832), branched cleanly off `main`.

## What lands
- **Kernel** `PtxFusedSoftmaxF32Kernel` — exact contiguous Ampere FP32 row softmax, two register-resident strategies:
  - *warp-row*: one warp per row, FP32x4/FP32x2 vector loads, butterfly-shuffle max+sum reductions, one coalesced vector store; **zero shared/local**.
  - *cta-row*: four-warp CTA reduction via 16-byte shared scratch for the (256,128) bucket; still one global load + one global store per element.
- **Backend** `CudaBackend.Softmax` fast-path (`TryDirectPtxSoftmax`), **gated OFF by default** (`AIDOTNET_DIRECT_PTX_SOFTMAX`).
- **62-cell coverage manifest** assigning every scoped softmax/sparsemax/logsumexp/masking API to a direct-PTX or existing-backend cell.
- **Benchmark** experiment + `--direct-ptx-softmax` flag and the resident PyTorch competitor script.
- **Tests** (4/4): register-resident pointer-only PTX structure, closed unpromoted shape domain, manifest completeness by reflection, and a driver correctness check that ran on the local RTX 3080 within **5e-5** of a double-precision oracle.

## Promotion status — fails closed
`IsPromotedShape => false`; production returns `softmax-performance-gate-not-met`. No promotion is claimed until three clean independent runs clear the `DirectPtxReleaseGate` (≥1.10x median vs the strongest competitor, ≤10% P95 regression, 0 managed/temp bytes, ≤5e-5 error, 0 local bytes). Admitted buckets: (256,128), (2048,64), (2048,128), (8192,128); everything else falls back with an exact reason.

## Fair-comparison notes (apples-to-apples)
The competitor script measures **four** resident PyTorch variants — eager, CUDA-graph, `max-autotune-no-cudagraphs` compile, and compile+graph — each against a double-precision softmax oracle, so our graph-captured kernel is compared against PyTorch's *graph-captured* kernel (symmetric launch-overhead amortization), NVIDIA-only (CPU MKL/OpenBLAS ineligible).

**Disclosed asymmetry:** the kernel uses approximate transcendentals (`ex2.approx.ftz` + `rcp.approx.ftz`, semantic `inference-forward-fast-exp2-rcp`, ~1e-5 error) whereas `torch.softmax` uses precise `expf`. Any promotion claim must state that the win is at a disclosed ~5e-5 tolerance, not bit-exact parity. Two protocol hardening items for the family (tracked for follow-up): **lock GPU clocks** (`nvidia-smi -lgc`) and **interleave** the C#/Python halves rather than running them sequentially, to prevent the thermal-drift artifact the pointwise blueprint already had to exclude a run for.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added an optimized direct-PTX FP32 row-wise softmax path for validated CUDA hardware and supported tensor shapes.
  * Added a benchmark command and help documentation for evaluating direct-PTX softmax performance.

* **Bug Fixes**
  * Improved benchmark argument forwarding and command handling.

* **Testing & Validation**
  * Added parity, numerical accuracy, resource, performance, and release-evidence checks.
  * Added benchmarking across multiple execution modes, including timing, throughput, memory, and error metrics.

* **Documentation**
  * Added rollout guidance, supported configurations, fallback behavior, and reproduction instructions.

### CODE REVIEW

**Comment by ooples** [comment]
Integrated the current p14 kernel foundation in 55040513, preserving the softmax-specific CLI, experiment gate, emitter, coverage manifest, and parity registration alongside the shared convolution/codegen/oracle infrastructure.

The cross-target validation exposed and fixed a real softmax issue before push: the public CudaBackend.Softmax route was still compile-time disabled on net471 even though the Direct PTX kernel now builds and runs there. The stale conditional is removed, so the existing promoted-geometry test observes all 34 dispatches on both targets. The softmax environment gate now also uses the same process-start snapshot contract as the rest of the resident launch path; test overrides remain dynamic.

Validation:
- benchmark project build: passed
- net10 softmax + shared foundation slice: 84/84 passed
- net471 softmax + standalone transfer-ordering slice: 15/15 passed
- exact previously failing net471 promoted public-route test: passed
- git diff checks: clean

No correctness, stability, architecture, or promotion gate was weakened.

**Comment by ooples** [comment]
Merged the final strengthened #909 foundation while preserving #868's production softmax route, net471 dispatch fix, and process-start feature-gate contract. Validation: Release net10.0 softmax + combined foundation suite 128/128; Release net471 softmax/transfer 15/15; Release benchmark build 0 errors. Branch clean; unresolved review threads 0.

**Comment by ooples** [comment]
Merged the final #909 foundation at 4baf6272. This branch already used the narrow artifact rule; package inventory remains 61 total resources (60 PTX artifacts), duplicates 0. Validation: DirectPTX 206/206, benchmark build 0 errors, and unresolved review threads 0.

**Comment by ooples** [comment]
Merged canonical scalar-parameter/offline-cubin foundation in 3ad72225. Revalidation: DirectPTX 206 passed with 1 intentional existing skip; benchmark project 0 errors (22 existing warnings); 60 packaged artifacts (58 cubins + 2 manifests), 0 duplicate logical names. Post-push review audit: 0 unresolved threads. Performance-oracle work remains separate.

**Comment by ooples** [comment]
Propagated the reviewed canonical offline-tool foundation in 6532d19b. Validation: benchmark build 0 errors; DirectPTX net10.0 206 passed + 1 existing runtime skip, 0 failures; attention ptxas manifest 15/15; attention nvdisasm SASS 15/15; embedded artifacts 60 = 58 cubins + 2 manifests with 0 duplicate hashes; worktree clean.

File: src/AiDotNet.Tensors/Engines/DirectGpu/CUDA/Ptx/PtxFusedSoftmaxF32Kernel.cs [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+
+/// <summary>
+/// Exact contiguous FP32 row-softmax specialization. A warp owns a row and
+/// retains every lane value from the sole global load through both reductions
+/// to the sole final global store.
+/// </summary>
+internal sealed class PtxFusedSoftmaxF32Kernel : IDisposable
+{
+    internal const string EntryPoint = "aidotnet_fused_softmax_f32";
+    internal const int DefaultBlockThreads = 512;
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/AiDotNet.Tensors/Engines/DirectGpu/CUDA/Ptx/PtxFusedSoftmaxF32Kernel.cs:20]: Verified against the current merged main: Softmax already has native routes on HIP, Metal, OpenCL, Vulkan, and WebGPU. DirectPtxSoftmaxCoverageManifest records those implementations, and SoftmaxCoverageManifest_TracksEveryMissingPeerDispatch requires every peer route to be native for CudaBackend.Softmax. This PR adds only a CUDA exact-shape optimization behind that existing cross-backend operation; the kernel remarks now make that scope explicit.

File: src/AiDotNet.Tensors/Engines/DirectGpu/CUDA/CudaBackend.cs [hunk 2]
```diff
@@ -3330,6 +3330,8 @@ public void Gelu(IGpuBuffer A, IGpuBuffer B, int size)
 
     public void Softmax(IGpuBuffer A, IGpuBuffer B, int batchSize, int features)
     {
+        if (TryDirectPtxSoftmax(A, B, batchSize, features))
+            return;
         // Fail-closed direct-PTX fast path (issue #840); returns false until GPU-promoted.
         if (TryDirectPtxSoftmax(A, B, batchSize, features)) return;
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/AiDotNet.Tensors/Engines/DirectGpu/CUDA/CudaBackend.cs]: Fixed in d294837f. CudaBackend.Softmax now makes exactly one TryDirectPtxSoftmax attempt and immediately falls back to LaunchSoftmaxKernel when the route declines.

File: src/AiDotNet.Tensors/Engines/DirectGpu/CUDA/Ptx/PtxFusedSoftmaxF32Kernel.cs [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+            runtime.ArchitectureFamily, rows, columns, blockThreads);
+        Ptx = EmitPtx(runtime.ComputeCapabilityMajor, runtime.ComputeCapabilityMinor,
+            rows, columns, blockThreads);
+        _module = runtime.LoadModule(Ptx);
+        _function = _module.GetFunction(EntryPoint, out DirectPtxFunctionInfo info);
+        int activeBlocks = _module.GetActiveBlocksPerMultiprocessor(_function, BlockThreads);
+        Blueprint.ResourceBudget.Validate(EntryPoint, info, BlockThreads, activeBlocks);
+        Audit = DirectPtxKernelAudit.Create(
+            Blueprint, runtime.DeviceFingerprint, Ptx, info,
+            BlockThreads, activeBlocks, _module);
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/AiDotNet.Tensors/Engines/DirectGpu/CUDA/Ptx/PtxFusedSoftmaxF32Kernel.cs]: Fixed in d294837f. PtxFusedSoftmaxF32Kernel now uses DirectPtxResourceInitialization.Complete, so function lookup, occupancy validation, resource validation, and audit creation all remain under temporary module ownership and cleanup cannot mask the primary failure.

File: tests/AiDotNet.Tensors.Benchmarks/Profiling/run-direct-ptx-softmax-evidence.ps1 [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            MaxError     = [double](@($m.Group | ForEach-Object { [double]$_.max_error } | Measure-Object -Maximum).Maximum)
+            Runs         = @($m.Group | Select-Object -ExpandProperty Process -Unique).Count
+        }
+    }
+
+    $direct = $methodAgg.Values | Where-Object { $_.Method.StartsWith($directMethodPrefix, [StringComparison]::Ordinal) } | Select-Object -First 1
+    $competitors = @($methodAgg.Values | Where-Object { -not $_.Method.StartsWith($directMethodPrefix, [StringComparison]::Ordinal) })
+    if (-not $direct -or $competitors.Count -eq 0) {
+        Write-Warning "[$shape] missing direct or competitor rows; skipping."
+        continue
```
[the remark this answers is not in the document]
Comment (ooples) [reply tests/AiDotNet.Tensors.Benchmarks/Profiling/run-direct-ptx-softmax-evidence.ps1:126]: Fixed in d294837f. The evidence runner now iterates the four required shapes, emits a HOLD manifest entry for missing shapes or missing direct/competitor rows, requires both sides to cover every requested process, and refuses a passing manifest unless all four entries exist.

File: tests/AiDotNet.Tensors.Benchmarks/Program.cs [hunk 5]
```diff
@@ -51,6 +51,25 @@ static void Main(string[] args)
...
+            Environment.ExitCode = DirectPtxAttentionOfflineCubinTool.Verify(args);
+            return;
+        }
+
+        if (args.Length > 0 && args[0] == "--audit-direct-ptx-attention-offline-sass")
+        {
+            Environment.ExitCode = DirectPtxAttentionOfflineCubinTool.AuditSass(args);
+            return;
+        }
+
```
[the remark this answers is not in the document]
Comment (ooples) [reply tests/AiDotNet.Tensors.Benchmarks/Program.cs]: Fixed in d294837f. All duplicate attention-offline cubin handler sets were removed, including the extra third copy created by the merge. One handler per command remains, and each strips the dispatch flag before calling the tool's operand-only API.

File: tests/AiDotNet.Tensors.Benchmarks/Program.cs [hunk 6]
```diff
@@ -1522,6 +1548,7 @@ static void Main(string[] args)
         Console.WriteLine("  --quick    : Run quick performance validation (default)");
         Console.WriteLine("  --full     : Run full BenchmarkDotNet suite (trigonometric)");
         Console.WriteLine("  --linalg   : Run linear algebra benchmarks vs MathNet.Numerics");
+        Console.WriteLine("  --direct-ptx-softmax [runs]: exact-shape FP32 row-softmax championship matrix");
```
[the remark this answers is not in the document]
Comment (ooples) [reply tests/AiDotNet.Tensors.Benchmarks/Program.cs]: Fixed in d294837f. Help now documents trailing operation filters, and the release-evidence runner invokes --direct-ptx-softmax without the obsolete numeric argument so every process emits the complete matrix.
